### PR TITLE
DDF-1610: Cometd Sample

### PR DESCRIPTION
**Pull request by oconnormi**
[Review on Reviewable](https://reviewable.io/reviews/codice/ddf/321)

### CODE REVIEW

**Comment by oconnormi** [comment]
@lessarderic
@figliold
@garrettfreibott
@andrewkfiedler


---







Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/321)

**Comment by lessarderic** [comment]
Reviewed 10 of 11 files at r1.
Review status: 10 of 11 files reviewed at latest revision, 8 unresolved discussions.

---

<sup>**[distribution/sdk/sample-cometd/bower.json, line 5 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/321#-K2hC8NO0YYsDs-v0Crq)** ([raw file](https://github.com/codice/ddf/blob/e5a196ff578f7e51a45867712e4e21d3f55f30a3/distribution/sdk/sample-cometd/bower.json#L5)):</sup>
Do we need the author property here? We usually avoid people's names and emails in the source code.


---

<sup>**[distribution/sdk/sample-cometd/Gruntfile.js, line 30 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/321#-K2hE5F-46e2-dmXphUF)** ([raw file](https://github.com/codice/ddf/blob/e5a196ff578f7e51a45867712e4e21d3f55f30a3/distribution/sdk/sample-cometd/Gruntfile.js#L30)):</sup>
Should we add `**` in those paths in case someone adds sub-directories in the future?


---

<sup>**[distribution/sdk/sample-cometd/package.json, line 17 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/321#-K2hER4aOTu9t03ZoW51)** ([raw file](https://github.com/codice/ddf/blob/e5a196ff578f7e51a45867712e4e21d3f55f30a3/distribution/sdk/sample-cometd/package.json#L17)):</sup>
Do we really need the `^` in the versions? Can't we just lock them in?


---

<sup>**[distribution/sdk/sample-cometd/README.md, line 7 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/321#-K2hFettKr3qqN3a8TEj)** ([raw file](https://github.com/codice/ddf/blob/e5a196ff578f7e51a45867712e4e21d3f55f30a3/distribution/sdk/sample-cometd/README.md#L7)):</sup>
Missing `cp` in the command.


---

<sup>**[distribution/sdk/sample-cometd/src/main/webapp/index.html, line 22 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/321#-K2hFwXB2Fp-vakJFf_J)** ([raw file](https://github.com/codice/ddf/blob/e5a196ff578f7e51a45867712e4e21d3f55f30a3/distribution/sdk/sample-cometd/src/main/webapp/index.html#L22)):</sup>
Interesting. Had never seen that `section` element before.


---

<sup>**[distribution/sdk/sample-cometd/src/main/webapp/index.html, line 33 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/321#-K2hGG2HJbkJMF6WZQD2)** ([raw file](https://github.com/codice/ddf/blob/e5a196ff578f7e51a45867712e4e21d3f55f30a3/distribution/sdk/sample-cometd/src/main/webapp/index.html#L33)):</sup>
Probably can delete that commented out line ;-)


---

<sup>**[distribution/sdk/sample-cometd/src/main/webapp/js/application.js, line 59 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/321#-K2hGr4k4EhdMmq93oj3)** ([raw file](https://github.com/codice/ddf/blob/e5a196ff578f7e51a45867712e4e21d3f55f30a3/distribution/sdk/sample-cometd/src/main/webapp/js/application.js#L59)):</sup>
Do you know if the `cometd.batch()` call is needed? I think I was able to get things working without it. If we keep it, maybe we should put a comment explaining why it's needed.


---

<sup>**[distribution/sdk/sample-cometd/src/main/webapp/js/application.js, line 70 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/321#-K2hHHFuFuNH_9PA-GRX)** ([raw file](https://github.com/codice/ddf/blob/e5a196ff578f7e51a45867712e4e21d3f55f30a3/distribution/sdk/sample-cometd/src/main/webapp/js/application.js#L70)):</sup>
Is the `publish` needed to flush out the messages already in the queue? If so, should we document that?


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/321)

File: distribution/sdk/sample-cometd/bower.json [hunk 1]
```diff
@@ -0,0 +1,22 @@
+{
+  "name": "ddf-cometd-browser-client",
+  "description": "Example of how to interact with ddf over cometd",
+  "main": "index.js",
+  "authors": [
+    "Michael O'Connor <michael.oconnor@connexta.com>"
```
Comment (stustison) [inline distribution/sdk/sample-cometd/bower.json]: Leave off the authors tag here. We've been slowly removing them all as we go along.

File: distribution/sdk/sample-cometd/package.json [hunk 2]
```diff
@@ -0,0 +1,24 @@
+{
```
Comment (stustison) [inline distribution/sdk/sample-cometd/package.json]: Might want to set "private": true

**Comment by oconnormi** [comment]
<sup>**[distribution/sdk/sample-cometd/src/main/webapp/js/application.js, line 70 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/321#-K2hHHFuFuNH_9PA-GRX)** ([raw file](https://github.com/codice/ddf/blob/e5a196ff578f7e51a45867712e4e21d3f55f30a3/distribution/sdk/sample-cometd/src/main/webapp/js/application.js#L70)):</sup>
Yep, publishing an empty message causes the server to send any persisted messages to the client


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/321)

**Comment by oconnormi** [comment]
<sup>**[distribution/sdk/sample-cometd/src/main/webapp/js/application.js, line 70 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/321#-K2hHHFuFuNH_9PA-GRX)** ([raw file](https://github.com/codice/ddf/blob/e5a196ff578f7e51a45867712e4e21d3f55f30a3/distribution/sdk/sample-cometd/src/main/webapp/js/application.js#L70)):</sup>
@lessarderic, now I just need to find the right place to document it, although I want to say that was in the documentation somewhere as I'm thinking that's how I knew about it in the first place


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/321)

**Comment by oconnormi** [comment]
<sup>**[distribution/sdk/sample-cometd/src/main/webapp/index.html, line 22 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/321#-K2hFwXB2Fp-vakJFf_J)** ([raw file](https://github.com/codice/ddf/blob/e5a196ff578f7e51a45867712e4e21d3f55f30a3/distribution/sdk/sample-cometd/src/main/webapp/index.html#L22)):</sup>
It's an html5 element, there are several new tags coming in that will replace the endless numbers of divs used by most pages in favor of a more semantic page structure


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/321)

**Comment by lessarderic** [comment]
Reviewed 1 of 11 files at r1.
Review status: 10 of 11 files reviewed at latest revision, 10 unresolved discussions.

---

<sup>**[distribution/sdk/sample-cometd/src/main/webapp/js/application.js, line 187 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/321#-K2hK29_sZSIBAgenwmy)** ([raw file](https://github.com/codice/ddf/blob/e5a196ff578f7e51a45867712e4e21d3f55f30a3/distribution/sdk/sample-cometd/src/main/webapp/js/application.js#L187)):</sup>
Should we define the base URL at the top of the file and use it to build the other URLs? Should we also put a note somewhere (README.md maybe?) saying that the sample assumes DDF is using the default hostname and port?


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/321)

**Comment by lessarderic** [comment]
Are you going to update the existing asciidocs in a separate PR?


---







Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/321)

**Comment by lessarderic** [comment]
<sup>**[distribution/sdk/sample-cometd/src/main/webapp/js/application.js, line 70 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/321#-K2hHHFuFuNH_9PA-GRX)** ([raw file](https://github.com/codice/ddf/blob/e5a196ff578f7e51a45867712e4e21d3f55f30a3/distribution/sdk/sample-cometd/src/main/webapp/js/application.js#L70)):</sup>
When I said "document", I only meant update the comment :-)


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/321)

**Comment by oconnormi** [comment]
<sup>**[distribution/sdk/sample-cometd/package.json, line 17 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/321#-K2hER4aOTu9t03ZoW51)** ([raw file](https://github.com/codice/ddf/blob/e5a196ff578f7e51a45867712e4e21d3f55f30a3/distribution/sdk/sample-cometd/package.json#L17)):</sup>
We could, though as dev dependencies its less important, also that was just the default way that npm adds them when using either `npm install --save foo` or `npm install --save-dev foo`


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/321)

**Comment by oconnormi** [comment]
<sup>**[distribution/sdk/sample-cometd/package.json, line 1 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/321#-K2gUtT1XoOCf-4DQ8LN-r1-1)** ([raw file](https://github.com/codice/ddf/blob/e5a196ff578f7e51a45867712e4e21d3f55f30a3/distribution/sdk/sample-cometd/package.json#L1)):</sup>
Good call


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/321)

**Comment by oconnormi** [comment]
<sup>**[distribution/sdk/sample-cometd/src/main/webapp/js/application.js, line 187 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/321#-K2hK29_sZSIBAgenwmy)** ([raw file](https://github.com/codice/ddf/blob/e5a196ff578f7e51a45867712e4e21d3f55f30a3/distribution/sdk/sample-cometd/src/main/webapp/js/application.js#L187)):</sup>
Alternatively, I could also make it a text box on the page and have it default to the current url. Thoughts?


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/321)

**Comment by oconnormi** [comment]
<sup>**[distribution/sdk/sample-cometd/src/main/webapp/js/application.js, line 59 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/321#-K2hGr4k4EhdMmq93oj3)** ([raw file](https://github.com/codice/ddf/blob/e5a196ff578f7e51a45867712e4e21d3f55f30a3/distribution/sdk/sample-cometd/src/main/webapp/js/application.js#L59)):</sup>
Added a comment


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/321)

**Comment by oconnormi** [comment]
<sup>**[distribution/sdk/sample-cometd/package.json, line 1 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/321#-K2gUtT1XoOCf-4DQ8LN-r1-1)** ([raw file](https://github.com/codice/ddf/blob/e5a196ff578f7e51a45867712e4e21d3f55f30a3/distribution/sdk/sample-cometd/package.json#L1)):</sup>
Done.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/321)

**Comment by lessarderic** [comment]
Reviewed 6 of 6 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[distribution/sdk/sample-cometd/package.json, line 17 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/321#-K2hER4aOTu9t03ZoW51)** ([raw file](https://github.com/codice/ddf/blob/e5a196ff578f7e51a45867712e4e21d3f55f30a3/distribution/sdk/sample-cometd/package.json#L17)):</sup>
:+1: 


---

<sup>**[distribution/sdk/sample-cometd/src/main/webapp/js/application.js, line 187 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/321#-K2hK29_sZSIBAgenwmy)** ([raw file](https://github.com/codice/ddf/blob/e5a196ff578f7e51a45867712e4e21d3f55f30a3/distribution/sdk/sample-cometd/src/main/webapp/js/application.js#L187)):</sup>
That'd be nicer for sure, but I'm not sure how much more time we want to spend improving this. Up to you. I'm fine either way.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/321)

**Comment by oconnormi** [comment]
I will add them to this PR, wasn't sure exactly where to put them though, I think it would be part of the Extending or Integrating doc under the ddf/catalog/ui/docs section. There doesn't seem to be much on either of those pages currently


---







Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/321)

**Comment by oconnormi** [comment]
<sup>**[distribution/sdk/sample-cometd/src/main/webapp/js/application.js, line 187 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/321#-K2hK29_sZSIBAgenwmy)** ([raw file](https://github.com/codice/ddf/blob/e5a196ff578f7e51a45867712e4e21d3f55f30a3/distribution/sdk/sample-cometd/src/main/webapp/js/application.js#L187)):</sup>
I just added a comment to the readme regarding using other urls, and moved the `var cometURL` up to the top of the `application.js` file


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/321)

**Comment by lessarderic** [comment]
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/321)

**Comment by lessarderic** [comment]
:+1: 


---

Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

<sup>**[catalog/ui/docs/src/main/resources/IntContents.adoc, line 9 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/321#-K2reYbhYnxyjpQVhwbp)** ([raw file](https://github.com/codice/ddf/blob/a079f004d145427ae00149866e5b33e56144d1e8/catalog/ui/docs/src/main/resources/IntContents.adoc#L9)):</sup>
I'd refer to the SDK for an example. Other than that lgtm.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/321)

**Comment by oconnormi** [comment]
<sup>**[catalog/ui/docs/src/main/resources/IntContents.adoc, line 9 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/321#-K2reYbhYnxyjpQVhwbp)** ([raw file](https://github.com/codice/ddf/blob/a079f004d145427ae00149866e5b33e56144d1e8/catalog/ui/docs/src/main/resources/IntContents.adoc#L9)):</sup>
Just added it


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/321)

**Comment by oconnormi** [comment]
Reviewed 4 of 11 files at r1, 5 of 6 files at r2.
Review status: 11 of 12 files reviewed at latest revision, 2 unresolved discussions.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/321)

**Comment by lessarderic** [comment]
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/321)

**Pull request merged**